Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -357,92 +377,174 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
-                throw error
+        case .headerField, .headerValue:
+            guard let headerStartIdx = self.state.headerStartIndex else {
+                return
             }
 
-            state.baseAddress = nil
+            self.mayDiscardDecodedBytes(readerIndex: headerStartIdx) {
+                // Reset the previous stored index that marks the start of the headers / trailers as we will adjust indices now.
+                self.state.headerStartIndex = nil
+
+                // We need to adjust the stored slice, currentNameIndex and also all the previous stored headers to reflect the new readerIndex after we discarded the bytes.
+                if let slice = self.state.slice {
+                    self.state.slice = (readerIndex: slice.readerIndex - headerStartIdx, slice.length)
+                }
+
+                func adjustedHeaderIndex(_ idx: HTTPHeaderIndex) -> HTTPHeaderIndex {
+                    return HTTPHeaderIndex(start: idx.start - headerStartIdx, length: idx.length)
+                }
+
+                if let idx = self.state.currentNameIndex {
+                    self.state.currentNameIndex = adjustedHeaderIndex(idx)
+                }
+
+                if !self.state.currentHeaders.isEmpty {
+                    self.state.currentHeaders = self.state.currentHeaders.map { HTTPHeader(name: adjustedHeaderIndex($0.name), value: adjustedHeaderIndex($0.value)) }
+                }
+            }
 
-            let httpError = parser.http_errno
-            if httpError != 0 {
-                self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
-                throw self.state.currentError!
+        case .status, .url:
+            assert(self.state.headerStartIndex == nil)
+            if let slice = self.state.slice {
+                self.mayDiscardDecodedBytes(readerIndex: slice.readerIndex) {
+                    // We discarded everything before the slice so the slice now starts at index 0.
+                    self.state.slice = (readerIndex: 0, slice.length)
+                }
             }
-            return result
-        }
-
-        if let slice = state.slice {
-            // If we have a slice, we need to preserve all of these bytes. To do that, we move the
-            // reader index to where the slice wants it, and then record how many readable bytes that leaves
-            // us with. Then, invalidate the reader index, as it's not stable across calls to decode()
-            // *anyway*, so we want to make sure we can see the bad value in debug errors.
-            buffer.moveReaderIndex(to: slice.readerIndex)
-            state.readerIndexAdjustment = buffer.readableBytes
-            state.slice = (-1, slice.length)
-            return .needMoreData
+        }
+    }
+
+    /// Will discard bytes till readerIndex if its needed and then call `fn`.
+    private func mayDiscardDecodedBytes(readerIndex: Int, _ fn: () -> Void) {
```

Comment:
I think that the "readerIndex" label could be improved: maybe "upTo" so that this now reads `mayDiscardDecodedBytes(upTo:)`.